remove uses of base_type_eq

kroening opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

kroening committed:
remove uses of base_type_eq

base_type_eq has been deprecated since 2019; a previous attempt to remove it
(#4056) has failed.

tautschnig commented on src/analyses/does_remove_const.cpp:
```diff
@@ -69,11 +68,11 @@ bool does_remove_constt::does_expr_lose_const(const exprt &expr) const
   {
     const typet &op_type=op.type();
-    if(base_type_eq(op_type, root_type, ns))
+    if(op_type == root_type)
```
With this change the `ns` class member should be removed in a follow-up commit.

kroening committed:
remove namespacet from does_remove_constt

kroening committed:
remove base_type_eq

The last uses have been removed, and base_type_eq has been deprecated since
2019; a previous attempt to remove it (#4056) has failed.

tautschnig commented on src/analyses/does_remove_const.cpp:
```diff
@@ -13,19 +13,15 @@ Author: Diffblue Ltd.
-  const namespacet &ns):
-    goto_program(goto_program),
-    ns(ns)
+does_remove_constt::does_remove_constt(const goto_programt &goto_program)
+  : goto_program(goto_program)
```
Thank you for adding this commit. The documentation above, however, also needs to be updated (removing lines 21 and 22) to keep doxygen happy.
